### PR TITLE
[tools/depends] bump expat 2.4.1

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
bump expat 2.4.1

## Motivation and context
dependency bump

@wsnipex another one thanks

A TODO here would be to switch over from autoconf to cmake for building expat. 

@Paxxi not sure where windows is regarding versions for expat, but noticed in the changelogs of expat 2.2.8 they renamed their cmake options
https://github.com/libexpat/libexpat/blob/a28238bdeebc087071777001245df1876a11f5ee/expat/Changes#L262-L279
They also deprecated Visual Studio <=14.0/2015 in 2.3.0 - https://github.com/libexpat/libexpat/blob/a28238bdeebc087071777001245df1876a11f5ee/expat/Changes#L114
May or may not be useful to you, just something i spotted.

## How has this been tested?
build osx

## What is the effect on users?
N/A

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
Going to need some advice on this one.

```
duplicate symbol '_testingAccountingGetCountBytesIndirect' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_testingAccountingGetCountBytesDirect' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_XML_SetBillionLaughsAttackProtectionMaximumAmplification' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_unsignedCharToPrintable' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_XML_SetBillionLaughsAttackProtectionActivationThreshold' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
ld: 5 duplicate symbols for architecture x86_64
```

All of these symbols were introduced in expat 2.4.0+, and essentially we are getting duplicate symbols from static build of python and then our use of expat in the kodi binary.

My solution was to introduce a dependency (expat-python) and then have expat then depend on it to insure the python build and use of python expat uses one set of symbols, and then the rest use renamed symbols.
Seems messy, but cant come up with a better way. Ideas greatly appreciated.

**Comment by wsnipex** [comment]
Are those warnings or errors? I'd assume ld issuing a warning, but silently using the first lib listed to link against.

**Comment by Paxxi** [comment]
Doesn't python want expat as a shared library _expat.pyd for the module?

**Comment by fuzzard** [comment]
> Are those warnings or errors? I'd assume ld issuing a warning, but silently using the first lib listed to link against.

They are errors on clang
```
ld: 6 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

As for the shared library, everything in the depends build for python modules are built static. So i guess if that was the required way, would have to rework everything in python

**Comment by wsnipex** [comment]
is the LDFLAG "-all_load" enabled? That would explain the error

**Comment by fuzzard** [comment]
Not that i can find.

**Comment by fuzzard** [comment]
> Doesn't python want expat as a shared library _expat.pyd for the module?

So with a quick workup, the reason python modules are handled the way they are currently (statically built into libpython.x.x.a) is android doesnt play nice with loading from default python locations (lib/pythonx.x/lib-dynload).
Would need to add more patches to make it work. 
I dont think at this time its worth going through that hassle

File: tools/depends/target/Makefile [hunk 1]
```diff
@@ -15,7 +15,8 @@ DEPENDS = \
 	taglib libusb libnfs freetype2 \
 	pythonmodule-pil pythonmodule-pycryptodome pythonmodule-setuptools \
 	libxslt ffmpeg crossguid libudfread \
-	libdvdread libdvdnav libdvdcss p8-platform flatbuffers dav1d xz
+	libdvdread libdvdnav libdvdcss p8-platform flatbuffers dav1d xz \
+  expat-python
```
Comment (garbear) [inline tools/depends/target/Makefile:19]: do we care about a little misalignment?

Comment (fuzzard) [reply tools/depends/target/Makefile:19]: I'll fix that up in the rebase if the approach is deemed acceptable

Comment (garbear) [reply tools/depends/target/Makefile:19]: I'm having flashbacks of all the times I've fixed rebase conflicts in this exact stanza. Please let's sort it and break into individual lines.

**Comment by fuzzard** [comment]
Heres my latest attempt. once again, opinions very much wanted.

This time, bring in a patch to set functions that cause duplicate symbols to be static functions.

The following 3 symbols had static removed due to test suite. For our purposes, they are internal, and we dont run test suite, so can either static or inline them.

- testingAccountingGetCountBytesDirect
- testingAccountingGetCountBytesIndirect
- unsignedCharToPrintable

_INTERNAL_trim_to_complete_utf8_characters is also an internal only function, and the symbol was unexposed in https://github.com/libexpat/libexpat/pull/457

That leaves the following 2 that ive just set to static.

- XML_SetBillionLaughsAttackProtectionMaximumAmplification
- XML_SetBillionLaughsAttackProtectionActivationThreshold

Local testing gets building working. In theory i think it should be ok, but would like other opinions.

**Comment by fuzzard** [comment]
Ok, round 3... haha.

The alternate i can think of, and as i now implement it, i think its the way to go.

Currently for target python3 we use system supplied expat (ie tools/depends/target/expat lib). Python supplies expat internally as an option and currently it is up to date with expat 2.41. As part of their pyexpat module, they redefine expat declarations in  https://github.com/python/cpython/blob/main/Modules/expat/pyexpatns.h

This has not been updated to support the new expat functions. If we simply patch this to include the "new" expat functions, and link with this instead of system expat, we then get different symbols in libpython3.9.a and libexpat.a when kodi is linked.

problem solved, patch in python is far simpler to maintain than the previous expat patch. The only potential downside is no longer using system expat, therefore if cpython upstream are sloppy with keeping it up to date, it may fall behind. I think at this point in time its a tradeoff worth taking.

Once again, let me know preferences for which approach

**Comment by wsnipex** [comment]
Even though it's not optimal to use internal expat in python, it does look like a sane solution for the time being for finally get expat updated.

**Review by wsnipex — APPROVED** [review]

**Pull request merged**